src/site/markdown/manual/cloud.md

```diff
@@ -118,76 +118,69 @@ being configured with a list of hosts and ports so high availability is not an i
-        <KeyValuePair key="kubernetes.namespaceId" value="${k8s:namespaceId:-}"/>
-        <KeyValuePair key="kubernetes.namespaceName" value="${k8s:namespaceName:-}"/>
-        <KeyValuePair key="kubernetes.podID" value="${k8s:podId:-}"/>
-        <KeyValuePair key="kubernetes.podIP" value="${k8s:podIp:-}"/>
-        <KeyValuePair key="kubernetes.podName" value="${k8s:podName:-}"/>
-        <KeyValuePair key="kubernetes.imageId" value="${k8s:imageId:-}"/>
-        <KeyValuePair key="kubernetes.imageName" value="${k8s:imageName:-}"/>
-      </GelfLayout>
+
+Log4j provides a multitude of JSON generating layouts. In particular, [JSON
+Template Layout](layouts.html#JSONTemplateLayout) allows full schema
+customization and bundles ELK-specific layouts by default, which makes it a
+great fit for the bill.
+
+    <Socket name="Logstash"
+            host="${sys:logstash.host}"
+            port="12345"
+            protocol="tcp"
+            bufferedIo="true">
+        <JsonTemplateLayout eventTemplateUri="classpath:EcsLayout.json">
+            <EventTemplateAdditionalFields>
+                <KeyValuePair key="containerId" value="${docker:containerId:-}"/>
+                <KeyValuePair key="application" value="$${lower:${spring:spring.application.name:-spring}}"/>
+                <KeyValuePair key="kubernetes.serviceAccountName" value="${k8s:accountName:-}"/>
+                <KeyValuePair key="kubernetes.containerId" value="${k8s:containerId:-}"/>
+                <KeyValuePair key="kubernetes.containerName" value="${k8s:containerName:-}"/>
+                <KeyValuePair key="kubernetes.host" value="${k8s:host:-}"/>
+                <KeyValuePair key="kubernetes.labels.app" value="${k8s:labels.app:-}"/>
+                <KeyValuePair key="kubernetes.labels.pod-template-hash" value="${k8s:labels.podTemplateHash:-}"/>
+                <KeyValuePair key="kubernetes.master_url" value="${k8s:masterUrl:-}"/>
+                <KeyValuePair key="kubernetes.namespaceId" value="${k8s:namespaceId:-}"/>
+                <KeyValuePair key="kubernetes.namespaceName" value="${k8s:namespaceName:-}"/>
+                <KeyValuePair key="kubernetes.podID" value="${k8s:podId:-}"/>
+                <KeyValuePair key="kubernetes.podIP" value="${k8s:podIp:-}"/>
+                <KeyValuePair key="kubernetes.podName" value="${k8s:podName:-}"/>
+                <KeyValuePair key="kubernetes.imageId" value="${k8s:imageId:-}"/>
+                <KeyValuePair key="kubernetes.imageName" value="${k8s:imageName:-}"/>
+            </EventTemplateAdditionalFields>
+        </JsonTemplateLayout>
     </Socket>
 
 ### Logstash Configuration
 
-    input {
-      gelf {
-        host => "localhost"
-        use_tcp => true
-        use_udp => false
-        port => 12222
-        type => "gelf"
-      }
-    }
+We will configure Logstash to listen on TCP port 12345 for payloads of type JSON
+and then forward these to (either console and/or) an Elasticsearch server.
```

Comment:
In my testing gelf has to be configured so that it uses null to delimit log events, otherwise events that contain exceptions get split into two log events. Did you do something magical to make the tcp input with JSON suddenly behave properly?